app/src/main/java/org/alphatilesapps/alphatiles/Validator.java

```diff
@@ -0,0 +1,992 @@
+            this.id = inID;
+            this.name = inName;
+            this.mimeType = inMimeType;
+        }
+
+        protected String getName() {
+            return this.name;
+        }
+
+        protected String getMimeType(){
+            return this.mimeType;
+        }
+
+        protected String getId() {
+            return this.id;
+        }
+    }
+
+    private class DriveFolder extends GoogleDriveItem{
+
+        private final ArrayList<GoogleDriveItem> folderContents = new ArrayList<>();
+
+        protected DriveFolder(String driveFolderId) throws IOException {
+            this(driveFolderId, (String) driveService.files().get(driveFolderId).get("name"));
+        }
+        protected DriveFolder(String driveFolderId, String inName) throws IOException {
+            super(driveFolderId, inName, "application/vnd.google-apps.folder");
+            String pageToken = null;
+            do {
+                FileList result = driveService.files().list()
+                        .setQ("parents in '" + driveFolderId + "' and trashed = false")
+                        .setSpaces("drive")
+                        .setFields("nextPageToken, files(id, name, mimeType)")
+                        .setPageToken(pageToken)
+                        .execute();
+
+                for (File file : result.getFiles()){
+
+                    if (file.getMimeType().equals("application/vnd.google-apps.spreadsheet")) {
+                        folderContents.add(new GoogleSheet(file.getId(), file.getName()));
+                    }
+                    else if (file.getMimeType().equals("application/vnd.google-apps.folder")) {
+                        folderContents.add(new DriveFolder(file.getId(), file.getName()));
+                    }
+                    else {
+                        folderContents.add(new GoogleDriveItem(file.getId(), file.getName(), file.getMimeType()));
+                    }
+                 }
+                pageToken = result.getNextPageToken();
+            } while (pageToken != null);
+        }
+
+        protected DriveFolder getFolderFromName(String inName) throws ValidatorException{
+            for (DriveFolder item : this.getContentsOfType(DriveFolder.class)){
+                if (item.getName().equals(inName)){
+                    return item;
+                }
+            }
+            throw new ValidatorException("was not able to find " + inName + " in the drive folder " + this.getName());
+        }
+
+        protected void checkFileNameAgainstList(ArrayList<String> toMatch){
+            PriorityQueue<String> LongestFirstToMatch = new PriorityQueue<>(toMatch);
+            for (GoogleDriveItem item : new ArrayList<>(folderContents)) {
+                boolean hasMatchingName = false;
+                for (String candidate : new ArrayList<>(LongestFirstToMatch)) {
+                    if (item.getName().startsWith(candidate)) {
```

Comment:
I don't completely understand this method, so maybe you can help clarify. The files need to exactly match the strings in the list toMatch (wordlist), not just start with that string. Does this method account for that? For example, "rice_stalk.png" will not work as the image name for "rice" in the wordlist.

Reply:
The version you were looking at did have some bugs, but when functioning properly it looks through the words from longest to shortest. Thus, assuming "rice_stalk" also corresponds to a word in wordlist, it would look for a match for that one first, and by the time it reaches "rice" the file rice_stalk.png would already have been removed from the folderContents list. The intention of this was so that people can tack on extra information to there fileNames without causing issues (rice_stalk_REAL.png), but as I am thinking about I see there could be a lot of unintended consequences (like mispelling a name and then getting confused about the errors that are thrown), so maybe it would be better to stick with exact matches.

Reply:
The code that loads the audio and images currently just takes the words in the wordlist (in the LWC) and appends the file extension to find the file, so exact matches are required unless we change the code also.